submissions/daniil-bodryagin/oop-exercise/index-oop.js

```diff
@@ -0,0 +1,180 @@
+import { print } from './js/lib.js';
+
+class Creature {
+  constructor ({name, gender, legs}) {
+    this.name = name;
+    this.gender = gender;
+    this.legs = legs;
+    this._friends = [];
+  }
+
+  set friends(listOfFriends) {
+    this._friends = listOfFriends;
+  }
+
+  get friends() {
+    return this._friends.map(({name}) => name).join(', ');
+  }
+
+  toString() {
+    const {constructor: {name: species}, name, gender, legs, friends} = this;
+    return [species, name, gender, legs, friends].join('; ');
+  }
+}
+
+class Human extends Creature {
+  constructor ({name, gender, legs = 2, hands = 2, phrase = 'Hello!'}) {
+    super({name, gender, legs});
+    this.phrase = phrase;
+    this.hands = hands;
+  }
+
+  toString() {
+    const {hands, phrase} = this;
+    const creatureString = super.toString();
+    const [species, name, gender, legs, friends] = creatureString.split('; ');
+    return [species, name, gender, legs, hands, phrase, friends].join('; ');
```

Comment:
I see what you are doing here.
In OOP paradigm child classes should be unaware of how things are built in any base class.
If we want to add yet another shared property to the base class, we will need to change `toString` implementation in many child classes. We do not want this.
You **do not** need to fix this now.

Reply:
Now I see that I implemented indeed not good solution.
I wanted to keep initial order of properties, but `hands` are to be inserted into the middle of presentation string. So the main question, how to find the place in this string? Parsing comes to mind as the more common approach but at he same time as very complicated. What would be better to do in such cases?

Reply:
Keep a map (list) of properties rather than converted string and use semantics like `insertPropertyAfter(property, target)` or `insertBefore...`
Do string conversion at the very last moment.

Reply:
Ok, thank you!

Reply:
When I added `phrase` to base class the `toString()` turned into simple `super.toString()` everywhere except in Human class. This is only class I leave as it is.

Reply:
@daniil-bodryagin Oh, I missed this one. You do not need to overload a method in child class if it's only function is to call an inherited version of the method. It is inherited by child classes and available out of the box by default.